src/Runner.Worker/Handlers/CompositeActionHandler.cs

```diff
@@ -34,7 +34,32 @@ public async Task RunAsync(ActionRunStage stage)
             Trace.Entering();
             ArgUtil.NotNull(ExecutionContext, nameof(ExecutionContext));
             ArgUtil.NotNull(Inputs, nameof(Inputs));
-            ArgUtil.NotNull(Data.Steps, nameof(Data.Steps));
+
+            List<Pipelines.ActionStep> steps;
+
+            if (stage == ActionRunStage.Pre)
+            {
+                ArgUtil.NotNull(Data.PreSteps, nameof(Data.PreSteps));
+                steps = Data.PreSteps;
+            } 
+            else if (stage == ActionRunStage.Post)
+            {
+                ArgUtil.NotNull(Data.PostSteps, nameof(Data.PostSteps));
+                steps = Data.PostSteps.ToList();
+                // Only register post steps for steps that actually ran
+                foreach (var step in steps)
+                {
+                    if (!ExecutionContext.Root.EmbeddedStepsWithPostRegistered.Contains(step.Id))
```

Comment:
👍 consistency w/ existing code `StepsWithPostRegistered`

wonder if we could change both to be more meaningful... i'm drawing a blank